### PR TITLE
Improved interface for partitions_add()

### LINKED ISSUE #72: Add helper function to create a biosboot partition
Add a function to add a biosboot partition with the `bios_boot` flag set, perhaps `mkbiosboot`?
As discussed in quattor/ncm-lib-blockdevices#49

### PR DESCRIPTION
**Pull request by jouvin**
Originally, `partitions_add` was only able to configure the size of a partition. The interface has been improved to received a dict of partition attributes rather than a number (partition size). For backward compatibility, the previous interface is still valid.

Partly fixes #72 (the complete fix requires the complementary improvement in template-library-core).

### CODE REVIEW

File: quattor/functions/filesystem.pan [hunk 1]
```diff
@@ -39,84 +20,92 @@ function num_of_harddisks = {
...
-            if (value(path)) {
-                device = i;
-            };
-        };
-        ok = next(dsk,i,v);
+  base = "/hardware/harddisks";
+  dsk = value(base);
+  device = "";
+  foreach (i;v;dsk) {
+    path = base+"/"+to_string(i)+"/boot";
```
Comment (stdweird) [inline quattor/functions/filesystem.pan]: use `format("%s/%s/boot", base, i)`

File: quattor/functions/filesystem.pan [hunk 2]
```diff
@@ -39,84 +20,92 @@ function num_of_harddisks = {
...
-        path = base+"/"+to_string(i)+"/boot";
-        if (exists(to_string(path))) {
-            if (value(path)) {
-                device = i;
-            };
-        };
-        ok = next(dsk,i,v);
+  base = "/hardware/harddisks";
+  dsk = value(base);
+  device = "";
```
Comment (stdweird) [inline quattor/functions/filesystem.pan]: not used?

File: quattor/functions/filesystem.pan [hunk 3]
```diff
@@ -39,84 +20,92 @@ function num_of_harddisks = {
...
-     };
-     # Validation stuff might be added here.
-     SELF;
+  function_name = 'lvs_add';
+  if (length (ARGV) != 2) {
+    error (function_name+": should get 2 arguments");
+  };
+
+  vg = ARGV[0];
+  lv=ARGV[1];
```
Comment (stdweird) [inline quattor/functions/filesystem.pan]: space around `=`

File: quattor/functions/filesystem.pan [hunk 4]
```diff
@@ -39,84 +20,92 @@ function num_of_harddisks = {
...
+    if (sz != -1) {
+        SELF[l]["size"] = sz;
+    };
+  };
+  # Validation stuff might be added here.
+  SELF;
 };
 
-# Adds a list of partitions to a disk. The third argument, if any, is
+# Adds a list of partitions to a disk. First argument is the holding device. Second argument
```
Comment (stdweird) [inline quattor/functions/filesystem.pan]: if you modify this, please also wrap it in `@documentation{}`

File: quattor/functions/filesystem.pan [hunk 5]
```diff
@@ -39,84 +20,92 @@ function num_of_harddisks = {
...
-     # Validation stuff might be added here.
-     SELF;
+  function_name = 'partitions_add';
+  if (length (ARGV) != 2 && length (ARGV) != 3) {
+    error (function_name+": requires 2 or 3 arguments");
+  };
+ 
+  pt=ARGV[1];
+  if (length (ARGV) == 3) {
+    ep=ARGV[2];
```
Comment (stdweird) [inline quattor/functions/filesystem.pan]: space around `=`

File: quattor/functions/filesystem.pan [hunk 6]
```diff
@@ -39,84 +20,92 @@ function num_of_harddisks = {
...
+  } else {
+    ep=undef;
```
Comment (stdweird) [inline quattor/functions/filesystem.pan]: space around `=`

File: quattor/functions/filesystem.pan [hunk 7]
```diff
@@ -39,84 +20,92 @@ function num_of_harddisks = {
...
+  };
+
+  foreach (p; params; pt) {
+    if (is_defined (ep)) {
+      ns=matches (p, ".*[^0-9]([0-9]+)$");
```
Comment (stdweird) [inline quattor/functions/filesystem.pan]: the `.*[^0-9]` is not needed

File: quattor/functions/filesystem.pan [hunk 8]
```diff
@@ -39,84 +20,92 @@ function num_of_harddisks = {
...
+      n=to_long (ns[1]);
```
Comment (stdweird) [inline quattor/functions/filesystem.pan]: space around `=`

File: quattor/functions/filesystem.pan [hunk 9]
```diff
@@ -39,84 +20,92 @@ function num_of_harddisks = {
...
+      if (n > 4) {
+        SELF[p]["type"] = "logical";
+      } else if (p == ep) {
```
Comment (stdweird) [inline quattor/functions/filesystem.pan]: can you test `== undef` (in case `ep` is `undef`)?

File: quattor/functions/filesystem.pan [hunk 10]
```diff
@@ -39,84 +20,92 @@ function num_of_harddisks = {
...
+        SELF[p]["type"] = "extended";
+      }; # Else, primary, which is the default
+    };
+    SELF[p]["holding_dev"] = ARGV[0];
+    if ( is_dict(params) ) {
+      part_params = params;
+    } else {
+      debug(format('%s: %s called with an implicit size argument for partition %s (%s)', OBJECT, function_name, p, to_string(params)));
+      part_params['size'] = params;
```
Comment (stdweird) [inline quattor/functions/filesystem.pan]: `part_params` is not pre-defined as an `dict()`. does this work?

**Comment by jouvin** [comment]
Sorry for not following up on this one (I missed the comments). But this is a requirement for https://github.com/quattor/template-library-standard/issues/85 that I would really like to see in 16.10.

File: quattor/functions/filesystem.pan (hunk 2 above)
Comment (jouvin) [inline quattor/functions/filesystem.pan]: This is not part of the function I modified but you are completely right...

File: quattor/functions/filesystem.pan (hunk 1 above)
Comment (jouvin) [inline quattor/functions/filesystem.pan]: idem

File: quattor/functions/filesystem.pan (hunk 3 above)
Comment (jouvin) [inline quattor/functions/filesystem.pan]: idem

File: quattor/functions/filesystem.pan (hunk 4 above)
Comment (jouvin) [inline quattor/functions/filesystem.pan]: Done.

File: quattor/functions/filesystem.pan (hunk 9 above)
Comment (jouvin) [inline quattor/functions/filesystem.pan]: Not sure to understand. This is wrapped into an `if ( is_defined(ep) ) ....`

**Comment by jouvin** [comment]
@stdweird comments addressed. Should be ready for merging...

File: quattor/functions/filesystem.pan (hunk 10 above)
Comment (stdweird) [inline quattor/functions/filesystem.pan]: @jouvin one left. does pan have autovivification?

Comment (jouvin) [inline quattor/functions/filesystem.pan]: No, I don't think so (except for SELF)... Good catch!

**Comment by jouvin** [comment]
Last comment addressed (https://github.com/quattor/template-library-core/pull/126#discussion_r86954796)!

**Comment by stdweird** [comment]
@jouvin thanks.

**Pull request merged**